### PR TITLE
Referrals api

### PR DESCRIPTION
**Pull request by ecotopian**
an example deep link would be:
ecosia://join/MANGO-MoUMgZ 

I reused our Loading Screen for Migration to claim a referral as it can be that both happen at the same time 😲

### CODE REVIEW

File: Client/Application/NavigationRouter.swift [hunk 1]
```diff
@@ -93,11 +96,11 @@ enum NavigationPath {
             return nil
         }
 
-        if urlString.starts(with: "\(scheme)://deep-link"), let deepURL = components.valueForQuery("url"), let link = DeepLink(urlString: deepURL.lowercased()) {
+        if urlString.starts(with: "ecosia://deep-link/trees"), let link = DeepLink(urlString: url.path) {
```
Comment (vauxhall) [inline Client/Application/NavigationRouter.swift:99]: @ecotopian Isn't it redundant to have a deep link named `deep-link` ?
Can it make more sense to have a descriptive name for what action is doing?

File: Client/Application/NavigationRouter.swift [hunk 2]
```diff
@@ -54,6 +55,8 @@ enum DeepLink {
             self = .homePanel(link)
         } else if component == "default-browser", let link = DefaultBrowserPath(rawValue: String(componentPath)) {
             self = .defaultBrowser(link)
+        } else if component == "trees" {
```
Comment (vauxhall) [inline Client/Application/NavigationRouter.swift:58]: Unfortunately when we migrated to our engine we lost our own implementation on deeplinks, but it seems like we could benefit on having an `enum` of possible `hosts` or `components` in a deeplink. At least for having a type safe comparison, if not for having an easier way to switching over all the variants.

File: Client/Application/NavigationRouter.swift [hunk 3]
```diff
@@ -210,6 +215,10 @@ enum NavigationPath {
         }
     }
 
+    private static func handleReferral(code: String, with bvc: BrowserViewController) {
```
Comment (vauxhall) [inline Client/Application/NavigationRouter.swift:218]: I feel this needs a bit of refactoring, I don't see a purpose for this method, the contents could be easily inlined on the caller removing some unnecessary complexity.

File: Client/Ecosia/Migration/LoadingScreen.swift [hunk 4]
```diff
@@ -133,4 +150,54 @@ final class LoadingScreen: UIViewController {
...
+        })
+        present(alert, animated: true)
+    }
+}
+
+extension Referrals.Error {
+    var title: String {
+        switch self {
+        case .badRequest:
+            return "Invalid referral link!"
```
Comment (vauxhall) [inline Client/Ecosia/Migration/LoadingScreen.swift:189]: Shouldn't errors be localized?

**Review by vauxhall — COMMENTED** [review]
Overall looks great, just some questions regarding the link.

File: Client/Application/NavigationRouter.swift (hunk 1 above)
Comment (ecotopian) [reply Client/Application/NavigationRouter.swift:99]: This builds on top of FFs Deeplink logic and my intention was to stay close to this. We haven't really agreed on an URL from product site yet. So we can still change this.

File: Client/Application/NavigationRouter.swift (hunk 2 above)
Comment (ecotopian) [reply Client/Application/NavigationRouter.swift:58]: yes indeed - I was about to wait how the final URL will look like and optimise for type safety afterwards

File: Client/Application/NavigationRouter.swift (hunk 1 above)
Comment (vauxhall) [reply Client/Application/NavigationRouter.swift:99]: Thanks for clarifying 👍

File: Client/Application/NavigationRouter.swift (hunk 3 above)
Comment (ecotopian) [reply Client/Application/NavigationRouter.swift]: removed :)

File: Client/Ecosia/Migration/LoadingScreen.swift (hunk 4 above)
Comment (ecotopian) [reply Client/Ecosia/Migration/LoadingScreen.swift]: now they are 👍 thanks for pointing out

File: Client/Application/NavigationRouter.swift (hunk 1 above)
Comment (ecotopian) [reply Client/Application/NavigationRouter.swift]: changed to `ecosia://join/code-123`

File: Client/Application/NavigationRouter.swift (hunk 2 above)
Comment (ecotopian) [reply Client/Application/NavigationRouter.swift]: I was giving this a second thought. Having only one custom deeplink, I tend to not refactor the FF deep-linking unless it's actually needed

**Comment by ecotopian** [comment]
@vauxhall can you please check again?

**Review by vauxhall — APPROVED** [review]
Looks great 🎉

**Pull request merged**